aten/src/ATen/native/TensorShape.cpp

```diff
@@ -1531,6 +1531,44 @@ Tensor flatten(const Tensor& self, DimnameList dims, Dimname out_dim) {
   return native::flatten(self, *dims.begin(), *(dims.end() - 1), out_dim);
 }
 
+Tensor unflatten(const Tensor& self, int64_t dim, IntArrayRef sizes, c10::optional<DimnameList> names) {
+  // unflatten is implemented only as a python method on tensor right now.
+  // The following asserts should be checked by the python method.
+  TORCH_INTERNAL_ASSERT(sizes.size() > 0);
+  TORCH_INTERNAL_ASSERT(!names || names->size() == sizes.size());
+
+  dim = maybe_wrap_dim(dim, self.dim());
+
+  int64_t numel = 1;
+  for (auto i : sizes) numel *= i;
+  TORCH_CHECK(numel == self.size(dim),
+      "unflatten: Provided sizes ", sizes, " don't multiply up to the size of dim ", 
+      dim, " (", self.names()[dim], ": ", self.size(dim), ") in Tensor", self.names());
+
+  auto shape = self.sizes().vec();
+  shape.erase(shape.begin() + dim);
+  shape.insert(shape.begin() + dim, sizes.begin(), sizes.end());
+
+  Tensor result;
+  {
+    NoNamesGuard guard;
+    result = self.reshape(shape);
```

Comment:
flatten uses reshape so I wanted to make it consistent with unflatten. Maybe we should change flatten to use view instead?

Reply:
The reason flatten uses `reshape` is because it is not always possible to flatten dimensions together. Take for instance:
```
x = torch.randn(2, 3).t()
x.view(6) # fails due to discontiguity
```
Now, we can always unflatten a single dimension into multiple dimensions by performing a view. Because of that, it doesn't matter if we call `reshape` or `view` -- the returned tensor will always be a view (reshape has two paths: if reshape can perform a view, then it will perform the view. Otherwise, it will perform a copy). 

Since `reshape` will always call `view` (and never perform a copy), we should make the semantics clear and do a `view` here.

Reply:
I see, thank you for clarifying.